### PR TITLE
Adopt Mutex where available

### PR DESCRIPTION
**Pull request by guoye-zhang**
Mutex should eliminate the extra allocation. Subclassing is the only way I found to work. I tried enums (enum case cannot be unavailable) and class-bound protocols (`isKnownUniquelyReferenced` doesn't work). @FranzBusch

### CODE REVIEW

File: Sources/HTTPTypes/HTTPFields.swift [hunk 1]
```diff
@@ -99,7 +109,40 @@ public struct HTTPFields: Sendable, Hashable {
         }
     }
 
-    private var _storage = _Storage()
+#if canImport(Synchronization)
+    @available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
+    private final class _StorageWithMutex: _Storage, @unchecked Sendable {
+        let mutex = Mutex<Void>(())
+
+        override func withLock<Result>(_ body: () throws -> Result) rethrows -> Result {
```
Comment (FranzBusch) [inline Sources/HTTPTypes/HTTPFields.swift:117]: Does this actually compile without warnings under strict concurrency? The closure that `Mutex` takes is `sending` whereas ours is not. This should produce a warning/error.

Comment (Lukasa) [reply Sources/HTTPTypes/HTTPFields.swift:117]: Needing dynamic dispatch is also a little gnarly.

Comment (guoye-zhang) [reply Sources/HTTPTypes/HTTPFields.swift]: I tried `.enableExperimentalFeature("StrictConcurrency")` and didn't get any errors, but added `sending` in 948fe6c4885ef0484237729c28d8df620da83ae4 to be safe.

Comment (guoye-zhang) [reply Sources/HTTPTypes/HTTPFields.swift]: I think dynamic dispatch is still better than an extra allocation for the lock?

Comment (guoye-zhang) [reply Sources/HTTPTypes/HTTPFields.swift:117]: Had to revert 948fe6c4885ef0484237729c28d8df620da83ae4 due to it breaking Swift 5.8/5.9

File: Sources/HTTPTypes/HTTPFields.swift [hunk 2]
```diff
@@ -99,7 +109,40 @@ public struct HTTPFields: Sendable, Hashable {
         }
     }
 
-    private var _storage = _Storage()
+    #if canImport(Synchronization)
+    @available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
```
Comment (guoye-zhang) [inline Sources/HTTPTypes/HTTPFields.swift:113]: Not sure how to fix this on Swift 5.9/5.10.

```
/__w/swift-http-types/swift-http-types/Sources/HTTPTypes/HTTPFields.swift:113:63: error: unrecognized platform name 'visionOS'
    @available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
```

**Review by Lukasa — APPROVED** [review]
I think this is honestly probably fine, we just need to be aware that one day we'll want to remove this.

**Review by ekinnear — APPROVED** [review]

**Pull request merged**